File: t/VB/PDFFormsDemo/PDFForms/PDFFormControls/ListFromField.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************

Imports System.Collections
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.Pdf
Imports System

Namespace PDFFormsDemo
   Public Class ListFormField
      Inherits FormFieldControl
      Implements IFormFieldList
      Public Sub New()
         _pdfListBox = New PDFListBox()
         Me.SetChildControl(_pdfListBox)
      End Sub

      Public Overrides Sub InitControl(formField As PDFFormField)
         MyBase.InitControl(formField)

         Dim listBox As ListBox = _pdfListBox.ListBox

         For Each content As String In formField.Contents
            _items.Add(content)
         Next

         UpdateItemsList(_items, listBox.Items)

         Dim ChoiceMultiSelect As Boolean = (formField.FieldFlags And PDFFormField.FieldFlagsChoiceMultiSelect) = PDFFormField.FieldFlagsChoiceMultiSelect
         Dim ChoiceSort As Boolean = (formField.FieldFlags And PDFFormField.FieldFlagsChoiceSort) = PDFFormField.FieldFlagsChoiceSort

         If ChoiceMultiSelect Then
            listBox.SelectionMode = SelectionMode.MultiSimple
         Else
            listBox.SelectionMode = SelectionMode.One
         End If

         If ChoiceSort Then
            Me.Sort()
         End If

         If formField.SelectedContents.Count > 0 Then
            If listBox.SelectionMode = SelectionMode.MultiSimple Then
               For Each content As String In formField.SelectedContents
                  Dim index As Integer = listBox.Items.IndexOf(content)
                  listBox.SetSelected(index, True)
               Next
            Else
               listBox.SelectedItem = formField.SelectedContents(0)
            End If
         End If
      End Sub

      Public Overrides Sub SetToolTip(toolTip As String)
         If FormFieldControl.FormFieldsToolTip IsNot Nothing Then
            FormFieldControl.FormFieldsToolTip.SetToolTip(_pdfListBox.ListBox, toolTip)
         End If

         MyBase.SetToolTip(toolTip)
      End Sub

      Public Overrides Sub SetContextMenuStrip()
         MyBase.SetContextMenuStrip()

         If FormFieldControl.FormFieldsContextMenu IsNot Nothing Then
            Dim listBox As ListBox = _pdfListBox.ListBox
            listBox.Tag = Me

            listBox.ContextMenuStrip = FormFieldControl.FormFieldsContextMenu
         End If
      End Sub

      Public Overrides Sub UpdateFormField(formField As PDFFormField)
         MyBase.UpdateFormField(formField)

         Dim listBox As ListBox = _pdfListBox.ListBox

         formField.Contents.Clear()
         For Each item As String In listBox.Items
            Dim itemContent As String = item
            formField.Contents.Add(itemContent)
         Next

         formField.SelectedContents.Clear()
         For Each item As String In listBox.SelectedItems
            Dim selectedItem As String = item
            formField.SelectedContents.Add(selectedItem)
         Next

         If listBox.SelectionMode = SelectionMode.MultiSimple Then
            formField.FieldFlags = formField.FieldFlags Or PDFFormField.FieldFlagsChoiceMultiSelect
         End If

         If _isSorted Then
            formField.FieldFlags = formField.FieldFlags Or PDFFormField.FieldFlagsChoiceSort
         End If
      End Sub

      Protected Overrides Sub OnPropertyChange()
         Me._pdfListBox.Enabled = Not _isReadOnly

         MyBase.OnPropertyChange()
      End Sub

      Public Sub Sort() Implements IFormFieldList.Sort
         _items.Sort()

         'Update ListBox items.
         UpdateItemsList(_items, _pdfListBox.ListBox.Items)

         _isSorted = True
      End Sub

#Region "Properties"

#Region "PDFListBox"

      Private _pdfListBox As PDFListBox
      Public ReadOnly Property PDFListBox() As PDFListBox
         Get
            Return Me._pdfListBox
         End Get
      End Property

#End Region

#Region "Items"

      Private _items As New List(Of String)()
      Public Property Items() As List(Of String) Implements IFormFieldList.Items
         Get
            Return _items
         End Get
         Set(value As List(Of String))
            If value IsNot Nothing Then
               UpdateItemsList(value, _items)
               UpdateItemsList(value, _pdfListBox.ListBox.Items)
               _isSorted = False
            End If
         End Set
      End Property

      Private Sub UpdateItemsList(newItems As List(Of String), target As IList)
         target.Clear()

         For i As Integer = 0 To newItems.Count - 1
            target.Add(newItems(i))
         Next
      End Sub

#End Region

#Region "SelectedIndex"

      Public Property SelectedIndex() As Integer Implements IFormFieldList.SelectedIndex
         Get
            Dim listBox As ListBox = _pdfListBox.ListBox
            Return listBox.SelectedIndex
         End Get
         Set(value As Integer)
            Dim listBox As ListBox = _pdfListBox.ListBox
            listBox.SelectedIndex = value
         End Set
      End Property

#End Region

#Region "SelectedIndices"

      Public Property SelectedIndices() As List(Of Integer) Implements IFormFieldList.SelectedIndices
         Get
            Dim listBox As ListBox = _pdfListBox.ListBox

            Dim selectedIndices__1 As New List(Of Integer)()
            For Each index As Integer In listBox.SelectedIndices
               selectedIndices__1.Add(index)
            Next

            Return selectedIndices__1
         End Get
         Set(value As List(Of Integer))
            Dim listBox As ListBox = _pdfListBox.ListBox
            listBox.ClearSelected()

            If value IsNot Nothing Then
               For Each index As Integer In value
                  listBox.SetSelected(index, True)
               Next
            End If
         End Set
      End Property

#End Region

#Region "IsSorted"

      Private _isSorted As Boolean = False
      Public Property IsSorted() As Boolean Implements IFormFieldList.IsSorted
         Get
            Return _isSorted
         End Get
         Set(value As Boolean)
            _isSorted = value

            If _isSorted Then
               Me.Sort()
            End If
         End Set
      End Property

#End Region

#End Region
   End Class

   Public Class PDFListBox
      Inherits Panel
      Implements IFormFieldControl
      Private _focus As Boolean = False
      Private _isBorderStyleUpdated As Boolean = False

      Public Sub New()
         _listBox = New ListBox()
         _listBox.BorderStyle = BorderStyle.None
         _listBox.Dock = DockStyle.Fill
         _listBox.IntegralHeight = False
         AddHandler _listBox.MouseEnter, AddressOf PDFListBox_MouseEnter
         AddHandler _listBox.MouseLeave, AddressOf PDFListBox_MouseLeave

         Me.Controls.Add(_listBox)
         AddHandler Me.MouseEnter, AddressOf PDFListBox_MouseEnter
         AddHandler Me.MouseLeave, AddressOf PDFListBox_MouseLeave
         AddHandler Me.Paint, AddressOf PDFListBox_Paint
      End Sub

      Private Sub PDFListBox_MouseEnter(sender As Object, e As EventArgs)
         _focus = True
         Me.Invalidate()
      End Sub

      Private Sub PDFListBox_MouseLeave(sender As Object, e As EventArgs)
         _focus = False
         Me.Invalidate()
      End Sub

      Private Sub PDFListBox_Paint(sender As Object, e As PaintEventArgs)
         If Not _isBorderStyleUpdated Then
            Me.FieldBorderStyle = _fieldBorderStyle
            _isBorderStyleUpdated = True
         End If

         ' Draw ListBox Border.
         FormFieldControl.DrawBorder(e.Graphics, Me, Me.ClientSize, _focus)
      End Sub

#Region "Properties"

#Region "ForeColor"

      Public Overrides Property ForeColor() As Color
         Get
            Return _listBox.ForeColor
         End Get
         Set(value As Color)
            _listBox.ForeColor = value
         End Set
      End Property

#End Region

#Region "BackgroundColor"

      Protected _backgroundColor As Color = SystemColors.Control
      Public Property BackgroundColor() As Color Implements IFormFieldControl.BackgroundColor
         Get
            Return _backgroundColor
         End Get
         Set(value As Color)
            _backgroundColor = value
            _listBox.BackColor = Me._backgroundColor
         End Set
      End Property

#End Region

#Region "BorderColor"

      Protected _borderColor As Color = Color.Black
      Public Property BorderColor() As Color Implements IFormFieldControl.BorderColor
         Get
            Return _borderColor
         End Get
         Set(value As Color)
            _borderColor = value
         End Set
      End Property

#End Region

#Region "BorderThickness"

      Private _borderThickness As Integer = 1
      Public Property BorderThickness() As Integer Implements IFormFieldControl.BorderThickness
         Get
            Return _borderThickness
         End Get
         Set(value As Integer)
            _borderThickness = value
         End Set
      End Property

#End Region

#Region "FieldBorderStyle"

      Protected _fieldBorderStyle As FieldBorderStyle = FieldBorderStyle.Solid
      Public Property FieldBorderStyle() As FieldBorderStyle Implements IFormFieldControl.FieldBorderStyle
         Get
            Return _fieldBorderStyle
         End Get
         Set(value As FieldBorderStyle)
            _fieldBorderStyle = value

            Select Case _fieldBorderStyle
               Case FieldBorderStyle.Solid, FieldBorderStyle.Dashed
                  Me.Padding = New Padding(_borderThickness)
                  Exit Select
               Case FieldBorderStyle.Beveled, FieldBorderStyle.Inset
                  Me.Padding = New Padding(_borderThickness * 2)
                  Exit Select
               Case FieldBorderStyle.Underlined
                  Me.Padding = New Padding(0, 0, 0, _borderThickness)
                  Exit Select
            End Select
         End Set
      End Property

#End Region

#Region "FocusedBorderColor"

      Protected _focusedBorderColor As Color = Color.FromArgb(255, 0, 153, 225)
      Public ReadOnly Property FocusedBorderColor() As Color Implements IFormFieldControl.FocusedBorderColor
         Get
            Return _focusedBorderColor
         End Get
      End Property

#End Region

#Region "IsFieldVisible"

      Private _isFieldVisible As Boolean = True
      Public Property IsFieldVisible() As Boolean Implements IFormFieldControl.IsFieldVisible
         Get
            Return _isFieldVisible
         End Get
         Set(value As Boolean)
            _listBox.Visible = value
            _isFieldVisible = value
         End Set
      End Property

#End Region

#Region "ListBox"

      Private _listBox As ListBox
      Public ReadOnly Property ListBox() As ListBox
         Get
            Return _listBox
         End Get
      End Property

#End Region

#End Region
   End Class
End Namespace

File: t/CS/PDFFormsDemo/PDFForms/PDFFormControls/ListFromField.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.Pdf;

namespace PDFFormsDemo
{
   public class ListFormField : FormFieldControl, IFormFieldList
   {
      public ListFormField()
      {
         _pdfListBox = new PDFListBox();
         this.SetChildControl(_pdfListBox);
      }

      public override void InitControl(PDFFormField formField)
      {
         base.InitControl(formField);

         ListBox listBox = _pdfListBox.ListBox;

         foreach (string content in formField.Contents)
         {
            _items.Add(content);
         }

         UpdateItemsList(_items, listBox.Items);

         bool ChoiceMultiSelect = (formField.FieldFlags & PDFFormField.FieldFlagsChoiceMultiSelect) == PDFFormField.FieldFlagsChoiceMultiSelect;
         bool ChoiceSort = (formField.FieldFlags & PDFFormField.FieldFlagsChoiceSort) == PDFFormField.FieldFlagsChoiceSort;

         if (ChoiceMultiSelect)
            listBox.SelectionMode = SelectionMode.MultiSimple;
         else
            listBox.SelectionMode = SelectionMode.One;

         if (ChoiceSort)
            this.Sort();

         if (formField.SelectedContents.Count > 0)
         {
            if (listBox.SelectionMode == SelectionMode.MultiSimple)
            {
               foreach (string content in formField.SelectedContents)
               {
                  int index = listBox.Items.IndexOf(content);
                  listBox.SetSelected(index, true);
               }
            }
            else
            {
               listBox.SelectedItem = formField.SelectedContents[0];
            }
         }
      }

      public override void SetToolTip(string toolTip)
      {
         if (FormFieldControl.FormFieldsToolTip != null)
         {
            FormFieldControl.FormFieldsToolTip.SetToolTip(_pdfListBox.ListBox, toolTip);
         }

         base.SetToolTip(toolTip);
      }

      public override void SetContextMenuStrip()
      {
         base.SetContextMenuStrip();

         if (FormFieldControl.FormFieldsContextMenu != null)
         {
            ListBox listBox = _pdfListBox.ListBox;
            listBox.Tag = this;

            listBox.ContextMenuStrip = FormFieldControl.FormFieldsContextMenu;
         }
      }

      public override void UpdateFormField(PDFFormField formField)
      {
         base.UpdateFormField(formField);

         ListBox listBox = _pdfListBox.ListBox;

         formField.Contents.Clear();
         foreach (string item in listBox.Items)
         {
            string itemContent = item;
            formField.Contents.Add(itemContent);
         }

         formField.SelectedContents.Clear();
         foreach (string item in listBox.SelectedItems)
         {
            string selectedItem = item;
            formField.SelectedContents.Add(selectedItem);
         }

         if (listBox.SelectionMode == SelectionMode.MultiSimple)
            formField.FieldFlags |= PDFFormField.FieldFlagsChoiceMultiSelect;

         if (_isSorted)
            formField.FieldFlags |= PDFFormField.FieldFlagsChoiceSort;
      }

      protected override void OnPropertyChange()
      {
         this._pdfListBox.Enabled = !_isReadOnly;

         base.OnPropertyChange();
      }

      public void Sort()
      {
         _items.Sort();

         //Update ListBox items.
         UpdateItemsList(_items, _pdfListBox.ListBox.Items);

         _isSorted = true;
      }

      #region Properties

      #region PDFListBox

      private PDFListBox _pdfListBox;
      public PDFListBox PDFListBox
      {
         get { return this._pdfListBox; }
      }

      #endregion //PDFListBox

      #region Items

      private List<string> _items = new List<string>();
      public List<string> Items
      {
         get
         {
            return _items;
         }
         set
         {
            if (value != null)
            {
               UpdateItemsList(value, _items);
               UpdateItemsList(value, _pdfListBox.ListBox.Items);
               _isSorted = false;
            }
         }
      }

      private void UpdateItemsList(List<string> newItems, IList target)
      {
         target.Clear();

         for (int i = 0; i < newItems.Count; i++)
         {
            target.Add(newItems[i]);
         }
      }

      #endregion //Items

      #region SelectedIndex

      public int SelectedIndex
      {
         get
         {
            ListBox listBox = _pdfListBox.ListBox;
            return listBox.SelectedIndex;
         }
         set
         {
            ListBox listBox = _pdfListBox.ListBox;
            listBox.SelectedIndex = value;
         }
      }

      #endregion //SelectedIndex

      #region SelectedIndices

      public List<int> SelectedIndices
      {
         get
         {
            ListBox listBox = _pdfListBox.ListBox;

            List<int> selectedIndices = new List<int>();
            foreach (int index in listBox.SelectedIndices)
            {
               selectedIndices.Add(index);
            }

            return selectedIndices;
         }
         set
         {
            ListBox listBox = _pdfListBox.ListBox;
            listBox.ClearSelected();

            if (value != null)
            {
               foreach (int index in value)
               {
                  listBox.SetSelected(index, true);
               }
            }
         }
      }

      #endregion //SelectedIndex

      #region IsSorted

      private bool _isSorted = false;
      public bool IsSorted
      {
         get { return _isSorted; }
         set
         {
            _isSorted = value;

            if (_isSorted)
            {
               this.Sort();
            }
         }
      }

      #endregion //IsSorted

      #endregion //Properties
   }

   public class PDFListBox : Panel, IFormFieldControl
   {
      private bool _focus = false;
      private bool _isBorderStyleUpdated = false;

      public PDFListBox()
      {
         _listBox = new ListBox();
         _listBox.BorderStyle = BorderStyle.None;
         _listBox.Dock = DockStyle.Fill;
         _listBox.IntegralHeight = false;
         _listBox.MouseEnter += PDFListBox_MouseEnter;
         _listBox.MouseLeave += PDFListBox_MouseLeave;

         this.Controls.Add(_listBox);
         this.MouseEnter += PDFListBox_MouseEnter;
         this.MouseLeave += PDFListBox_MouseLeave;
         this.Paint += PDFListBox_Paint;
      }

      private void PDFListBox_MouseEnter(object sender, EventArgs e)
      {
         _focus = true;
         this.Invalidate();
      }

      private void PDFListBox_MouseLeave(object sender, EventArgs e)
      {
         _focus = false;
         this.Invalidate();
      }

      private void PDFListBox_Paint(object sender, PaintEventArgs e)
      {
         if (!_isBorderStyleUpdated)
         {
            this.FieldBorderStyle = _fieldBorderStyle;
            _isBorderStyleUpdated = true;
         }

         // Draw ListBox Border.
         FormFieldControl.DrawBorder(e.Graphics, this, this.ClientSize, _focus);
      }

      #region Properties

      #region ForeColor

      public override Color ForeColor
      {
         get
         {
            return _listBox.ForeColor;
         }
         set
         {
            _listBox.ForeColor = value;
         }
      }

      #endregion //ForeColor

      #region BackgroundColor

      protected Color _backgroundColor = SystemColors.Control;
      public Color BackgroundColor
      {
         get
         {
            return _backgroundColor;
         }
         set
         {
            _backgroundColor = value;
            _listBox.BackColor = this._backgroundColor;
         }
      }

      #endregion //BackgroundColor

      #region BorderColor

      protected Color _borderColor = Color.Black;
      public Color BorderColor
      {
         get
         {
            return _borderColor;
         }
         set
         {
            _borderColor = value;
         }
      }

      #endregion //BorderColor

      #region BorderThickness

      private int _borderThickness = 1;
      public int BorderThickness
      {
         get
         {
            return _borderThickness;
         }
         set
         {
            _borderThickness = value;
         }
      }

      #endregion //BorderThickness

      #region FieldBorderStyle

      protected FieldBorderStyle _fieldBorderStyle = FieldBorderStyle.Solid;
      public FieldBorderStyle FieldBorderStyle
      {
         get
         {
            return _fieldBorderStyle;
         }
         set
         {
            _fieldBorderStyle = value;

            switch (_fieldBorderStyle)
            {
               case FieldBorderStyle.Solid:
               case FieldBorderStyle.Dashed:
                  this.Padding = new Padding(_borderThickness);
                  break;
               case FieldBorderStyle.Beveled:
               case FieldBorderStyle.Inset:
                  this.Padding = new Padding(_borderThickness * 2);
                  break;
               case FieldBorderStyle.Underlined:
                  this.Padding = new Padding(0, 0, 0, _borderThickness);
                  break;
            }
         }
      }

      #endregion //FieldBorderStyle

      #region FocusedBorderColor

      protected Color _focusedBorderColor = Color.FromArgb(255, 0, 153, 225);
      public Color FocusedBorderColor
      {
         get
         {
            return _focusedBorderColor;
         }
      }

      #endregion //FocusedBorderColor

      #region IsFieldVisible

      private bool _isFieldVisible = true;
      public bool IsFieldVisible
      {
         get
         {
            return _isFieldVisible;
         }
         set
         {
            _listBox.Visible = value;
            _isFieldVisible = value;
         }
      }

      #endregion //IsFieldVisible

      #region ListBox

      private ListBox _listBox;
      public ListBox ListBox
      {
         get { return _listBox; }
      }

      #endregion //ListBox

      #endregion //Properties
   }
}
